### PR TITLE
GCP-231: feat(api): add GCP Workload Identity Federation types and validation

### PR DESCRIPTION
**Pull request by apahim**
## What this PR does / why we need it:

  This PR introduces Workload Identity Federation (WIF) API types for the GCP platform in HyperShift. These changes lay the foundation for secure, keyless
   authentication between hosted control planes and GCP services.

  The PR adds:

  1. **GCP Workload Identity Federation API Types** (`feat(api)`):
     - `GCPWorkloadIdentityConfig`: Core structure for WIF configuration including service account references and project metadata
     - `GCPResourceLabel`: RFC1035-compliant labels for Compute Engine resources
     - Comprehensive validation rules following GCP specifications:
       - Project ID and region validation
       - Resource labels (RFC1035 format, no underscores)
       - Service account email validation with hardened regex patterns
       - Project number validation
       - Network and subnet name validation
       - Reserved prefix validation (blocks 'goog' prefix per GCP standards)
     - New condition type `ValidWorkloadIdentityConfiguration` for platform validation
     - Complete unit test coverage for validation rules

  2. **Client Apply Configurations** (`feat(client)`):
     - Generated apply configurations for declarative client operations with server-side apply
     - Support for `GCPPlatformSpec`, `GCPResourceLabel`, `GCPServiceAccountsRef`, and `GCPWorkloadIdentityConfig`

  All changes include:
  - Updated CRD manifests (feature-gated and standard)
  - Generated deepcopy functions
  - API documentation updates
  - Vendored API updates

  ## Which issue(s) this PR fixes:

  Fixes [GCP-231](https://issues.redhat.com//browse/GCP-231)

  ## Special notes for your reviewer:

  - This PR only adds the API layer for WIF - no controller implementation yet
  - All validation rules follow official GCP naming conventions and constraints
  - Resource labels specifically follow RFC1035 (used by Compute Engine) which differs from standard Kubernetes labels (RFC1123)
  - The `GCPWorkloadIdentityConfig` structure is designed to support both user-provided and HyperShift-managed service accounts
  - Validation tests cover edge cases including reserved prefixes, special characters, and length constraints

  ## Checklist:
  - [x] Subject and description added to both, commit and PR.
  - [x] Relevant issues have been referenced.
  - [ ] This change includes docs. (API docs auto-generated)
  - [x] This change includes unit tests.

### CODE REVIEW

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml [hunk 1]
```diff
@@ -5703,21 +5718,172 @@ spec:
...
+                        type: object
+                        x-kubernetes-validations:
+                        - message: WorkloadIdentity is immutable
+                          rule: self == oldSelf
                     required:
                     - networkConfig
                     - project
                     - region
+                    - workloadIdentity
                     type: object
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml:5917]: There are no existing GCPs HCPs

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml [hunk 2]
```diff
@@ -5703,21 +5718,172 @@ spec:
...
+                                  that manage NodePool infrastructure (VMs, networks, disks, etc.).
+                                  This GSA needs compute.*, network.*, and storage.* permissions.
+                                  Format: service-account-name@project-id.iam.gserviceaccount.com
+                                maxLength: 100
+                                minLength: 1
+                                pattern: ^[a-z][a-z0-9-]{4,28}[a-z0-9]@[a-z][a-z0-9-]{4,28}[a-z0-9]\.iam\.gserviceaccount\.com$
+                                type: string
+                                x-kubernetes-validations:
+                                - message: NodePool email is immutable
+                                  rule: self == oldSelf
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml]: this is intentional, no legacy accounts.

File: api/hypershift/v1beta1/gcp.go [hunk 3]
```diff
@@ -81,4 +127,110 @@ type GCPPlatformSpec struct {
...
+	//   - Workload Identity Pool and Provider must exist in the GCP project
+	//   - Provider must be configured with audience mapping for OpenShift SA tokens
+	//   - Target Google Service Account must have roles/iam.workloadIdentityUser
+	//     granted to the workload pool principal (e.g., principal://iam.googleapis.com/projects/PROJECT_NUMBER/locations/global/workloadIdentityPools/POOL_ID/subject/system:serviceaccount:kube-system:capi-gcp-controller-manager)
+	//   - Attribute mappings on the provider should include google.subject for token subject verification
+	//
+	// +required
+	// +immutable
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="WorkloadIdentity is immutable"
+	WorkloadIdentity GCPWorkloadIdentityConfig `json:"workloadIdentity"`
```
[the remark this answers is not in the document]
Comment (apahim) [reply api/hypershift/v1beta1/gcp.go]: this is fine, we are still developing GCP support behind a feature flag.

File: api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml [hunk 4]
```diff
@@ -4784,21 +4799,172 @@ spec:
...
+                        required:
+                        - poolID
+                        - projectNumber
+                        - providerID
+                        - serviceAccountsRef
+                        type: object
+                        x-kubernetes-validations:
+                        - message: WorkloadIdentity is immutable
+                          rule: self == oldSelf
                     required:
```
[the remark this answers is not in the document]
Comment (apahim) [reply api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml:4993]: this is fine, we are still developing GCP support behind a feature flag.

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedclusters-CustomNoUpgrade.crd.yaml [hunk 5]
```diff
@@ -5811,21 +5826,172 @@ spec:
...
+                              minLength: 0
+                              pattern: ^$|^[0-9a-z]([0-9a-z-]{0,61}[0-9a-z])?$
+                              type: string
+                          required:
+                          - key
+                          - value
+                          type: object
+                        maxItems: 64
+                        type: array
+                      workloadIdentity:
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedclusters-CustomNoUpgrade.crd.yaml:5890]: This is intentional, to acoid confusion.

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedclusters-CustomNoUpgrade.crd.yaml [hunk 6]
```diff
@@ -5811,21 +5826,172 @@ spec:
...
+                        type: object
+                        x-kubernetes-validations:
+                        - message: WorkloadIdentity is immutable
+                          rule: self == oldSelf
                     required:
                     - networkConfig
                     - project
                     - region
+                    - workloadIdentity
                     type: object
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedclusters-CustomNoUpgrade.crd.yaml:6025]: this is intentional, there are no existing clusters and we are still developing the feature under a feature flag.

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml]: no this is fine, no need to overcomplicate it.

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-TechPreviewNoUpgrade.crd.yaml [hunk 7]
```diff
@@ -5414,21 +5429,172 @@ spec:
...
+                              maxLength: 63
+                              minLength: 0
+                              pattern: ^$|^[0-9a-z]([0-9a-z-]{0,61}[0-9a-z])?$
+                              type: string
+                          required:
+                          - key
+                          - value
+                          type: object
+                        maxItems: 64
+                        type: array
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-TechPreviewNoUpgrade.crd.yaml:5489]: I will change to maxItems: 64, but list-map semantics (+listType=map) - this approach is discouraged

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml [hunk 8]
```diff
@@ -5703,21 +5718,172 @@ spec:
...
+                              minLength: 0
+                              pattern: ^$|^[0-9a-z]([0-9a-z-]{0,61}[0-9a-z])?$
+                              type: string
+                          required:
+                          - key
+                          - value
+                          type: object
+                        maxItems: 64
+                        type: array
+                      workloadIdentity:
```
Comment (apahim) [inline cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml:5782]: I will change to maxItems: 64, but list-map semantics (+listType=map) - this approach is discouraged

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml [hunk 9]
```diff
@@ -5703,21 +5718,172 @@ spec:
...
+                            minLength: 4
+                            pattern: ^[a-z0-9]([a-z0-9-]{2,30}[a-z0-9])$
+                            type: string
+                            x-kubernetes-validations:
+                            - message: Provider ID cannot start with reserved prefix
+                                'gcp-'
+                              rule: '!self.startsWith(''gcp-'')'
+                            - message: Provider ID is immutable
+                              rule: self == oldSelf
+                          serviceAccountsRef:
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml]: will do

File: docs/content/reference/api.md [hunk 10]
```diff
@@ -6153,7 +6260,127 @@ string
 </em>
 </td>
 <td>
-<p>name is the name of the GCP resource</p>
+<p>name is the name of the GCP resource.
+Must conform to GCP resource naming standards: lowercase letters, numbers, and hyphens only.
+Must start with lowercase letter and end with lowercase letter or number, max 63 characters.
+Pattern: ^<a href="[-a-z0-9]*[a-z0-9]">a-z</a>?$ per GCP naming requirements.
+See <a href="https://cloud.google.com/compute/docs/naming-resources">https://cloud.google.com/compute/docs/naming-resources</a> for details.</p>
+</td>
```
[the remark this answers is not in the document]
Comment (apahim) [reply docs/content/reference/api.md:6304]: will do

File: docs/content/reference/api.md [hunk 11]
```diff
@@ -6153,7 +6260,127 @@ string
...
+<td>
+<p>nodePoolEmail is the Google Service Account email for CAPG controllers
+that manage NodePool infrastructure (VMs, networks, disks, etc.).
+This GSA needs compute.<em>, network.</em>, and storage.* permissions.
+Format: service-account-name@project-id.iam.gserviceaccount.com</p>
+</td>
+</tr>
+</tbody>
+</table>
```
[the remark this answers is not in the document]
Comment (apahim) [reply docs/content/reference/api.md:6366]: will do

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedclusters-TechPreviewNoUpgrade.crd.yaml [hunk 12]
```diff
@@ -5522,21 +5537,177 @@ spec:
...
+                        type: object
+                        x-kubernetes-validations:
+                        - message: WorkloadIdentity is immutable
+                          rule: self == oldSelf
                     required:
                     - networkConfig
                     - project
                     - region
+                    - workloadIdentity
                     type: object
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedclusters-TechPreviewNoUpgrade.crd.yaml:5736]: The workloadIdentity field appearing only in TechPreviewNoUpgrade and CustomNoUpgrade variants is correct and intentional.

  Rationale:
  - The entire GCP platform is gated behind the GCPPlatform feature gate
  - This feature gate is disabled in the Default (GA) variant (see api/hypershift/v1beta1/featuregates/featureGate-Hypershift-Default.yaml:34)
  - It's enabled in TechPreviewNoUpgrade and CustomNoUpgrade variants (see featureGate-Hypershift-TechPreviewNoUpgrade.yaml:44)
  - Therefore, workloadIdentity correctly appears only where GCP platform support exists

  This follows the same pattern as other tech-preview features like OpenStack and AutoNodeKarpenter.

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedclusters-TechPreviewNoUpgrade.crd.yaml [hunk 13]
```diff
@@ -5522,21 +5537,184 @@ spec:
...
+
+                              This is a user-provided value obtained from GCP (found in GCP Console or via `gcloud projects describe PROJECT_ID`).
+                              Also available in the output of `hypershift infra create gcp`.
+                            maxLength: 25
+                            minLength: 1
+                            pattern: ^[0-9]+$
+                            type: string
+                            x-kubernetes-validations:
+                            - message: Project number is immutable
+                              rule: self == oldSelf
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedclusters-TechPreviewNoUpgrade.crd.yaml:5650]: is that a standard or a hard limit in the google api?

[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedclusters-TechPreviewNoUpgrade.crd.yaml:5650]: let's leave as is.

File: api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedcontrolplanes.hypershift.openshift.io/GCPPlatform.yaml [hunk 14]
```diff
@@ -4676,21 +4691,205 @@ spec:
...
+                              maxLength: 63
+                              minLength: 0
+                              pattern: ^$|^[0-9a-z]([0-9a-z-]{0,61}[0-9a-z])?$
+                              type: string
+                          required:
+                          - key
+                          - value
+                          type: object
+                        maxItems: 60
+                        type: array
```
[the remark this answers is not in the document]
Comment (apahim) [reply api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedcontrolplanes.hypershift.openshift.io/GCPPlatform.yaml:4751]: nope, we discussed this before.

File: docs/content/reference/api.md [hunk 15]
```diff
@@ -6153,7 +6260,157 @@ string
...
+that manage NodePool infrastructure (VMs, networks, disks, etc.).
+This GSA requires the following IAM roles:
+- roles/compute.instanceAdmin.v1 (Compute Instance Admin v1)
+- roles/compute.networkAdmin (Compute Network Admin)
+See cmd/infra/gcp/iam-bindings.json for the authoritative role definitions.
+Format: service-account-name@project-id.iam.gserviceaccount.com</p>
+<p>This is a user-provided value referencing a pre-created Google Service Account.
+Typically obtained from the output of <code>hypershift infra create gcp</code> which creates
+the required service accounts with appropriate IAM roles and WIF bindings.</p>
+</td>
```
[the remark this answers is not in the document]
Comment (apahim) [reply docs/content/reference/api.md]: this is fine, we set the WorkloadIdentityUser

File: api/hypershift/v1beta1/gcp.go [hunk 16]
```diff
@@ -4,14 +4,55 @@ package v1beta1
...
+	// - If non-empty, must start with a lowercase letter or digit
+	// - If non-empty, cannot end with a hyphen
+	// Note: Other GCP services may allow a broader character set (e.g., underscores); consult service-specific documentation.
+	// See https://cloud.google.com/compute/docs/labeling-resources for Compute Engine label requirements.
+	//
+	// +required
+	// +kubebuilder:validation:MinLength=0
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:Pattern=`^$|^[0-9a-z]([0-9a-z-]{0,61}[0-9a-z])?$`
+	Value string `json:"value"`
```
Comment (csrwng) [inline api/hypershift/v1beta1/gcp.go]: Same comment as in the other PR... if you want to allow empty string as a valid value and make this required, then it needs to be a pointer.

File: api/hypershift/v1beta1/gcp_validation_test.go [hunk 17]
```diff
@@ -0,0 +1,101 @@
+package v1beta1
```
Comment (csrwng) [inline api/hypershift/v1beta1/gcp_validation_test.go]: See https://github.com/openshift/hypershift/pull/7251#discussion_r2562843227

File: api/hypershift/v1beta1/gcp.go [hunk 18]
```diff
@@ -81,4 +127,122 @@ type GCPPlatformSpec struct {
...
+	// Typically obtained from the output of `hypershift infra create gcp` which creates
+	// the required service accounts with appropriate IAM roles and WIF bindings.
+	//
+	// +required
+	// +immutable
+	// +kubebuilder:validation:Pattern=`^[a-z][a-z0-9-]{4,28}[a-z0-9]@[a-z][a-z0-9-]{4,28}[a-z0-9]\.iam\.gserviceaccount\.com$`
+	// +kubebuilder:validation:MinLength=38
+	// +kubebuilder:validation:MaxLength=100
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="NodePool email is immutable"
+	NodePoolEmail string `json:"nodePoolEmail"`
```
Comment (csrwng) [inline api/hypershift/v1beta1/gcp.go]: See https://github.com/openshift/hypershift/pull/7251#discussion_r2562863914

**Review by csrwng — COMMENTED** [review]
Same comments as in the original PR
https://github.com/openshift/hypershift/pull/7251

File: api/hypershift/v1beta1/gcp.go (hunk 16 above)
Comment (apahim) [reply api/hypershift/v1beta1/gcp.go]: yeah this was coderabbit requesting empty strings because it's allowd in google API, but will probably fail later. I will not allow empty strings.

Comment (apahim) [reply api/hypershift/v1beta1/gcp.go]: well, empty strings are fine, I will use the pointer

File: api/hypershift/v1beta1/gcp.go [hunk 19]
```diff
@@ -81,4 +127,142 @@ type GCPPlatformSpec struct {
...
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Provider ID is immutable"
+	ProviderID string `json:"providerID"`
+
+	// serviceAccountsRef contains references to various Google Service Accounts
+	// required to enable integrations for different controllers and operators.
+	// This follows the AWS pattern of having different roles for different purposes.
+	//
+	// +required
+	// +immutable
+	ServiceAccountsRef GCPServiceAccountsRef `json:"serviceAccountsRef"`
```
Comment (muraee) [inline api/hypershift/v1beta1/gcp.go]: I think `serviceAccountsEmails` is more clear, and then for the sub fields, as Cesar suggested, we can change it to be just `nodePool` and `controlPlane`

File: api/hypershift/v1beta1/gcp.go (hunk 16 above)
Comment (muraee) [reply api/hypershift/v1beta1/gcp.go]: why not make it optional if empty strings are allowed?

File: api/hypershift/v1beta1/gcp.go (hunk 19 above)
Comment (apahim) [reply api/hypershift/v1beta1/gcp.go]: sure

File: api/hypershift/v1beta1/gcp.go [hunk 20]
```diff
@@ -4,14 +4,55 @@ package v1beta1
...
+	// - If non-empty, cannot end with a hyphen
+	// Note: Other GCP services may allow a broader character set (e.g., underscores); consult service-specific documentation.
+	// See https://cloud.google.com/compute/docs/labeling-resources for Compute Engine label requirements.
+	//
+	// +required
+	// +kubebuilder:validation:MinLength=0
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:Pattern=`^$|^[0-9a-z]([0-9a-z-]{0,61}[0-9a-z])?$`
+	Value *string `json:"value,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (apahim) [reply api/hypershift/v1beta1/gcp.go:50]: *string makes it allowed as empty string, but it's not optional

File: api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml [hunk 21]
```diff
@@ -4784,21 +4799,205 @@ spec:
...
+                              maxLength: 63
+                              minLength: 0
+                              pattern: ^$|^[0-9a-z]([0-9a-z-]{0,61}[0-9a-z])?$
+                              type: string
+                          required:
+                          - key
+                          - value
+                          type: object
+                        maxItems: 60
+                        type: array
```
[the remark this answers is not in the document]
Comment (apahim) [reply api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml:4859]: no, this was rejected before, won't do.

File: api/go.mod [hunk 22]
```diff
@@ -7,6 +7,7 @@ require (
 	github.com/openshift/api v0.0.0-20250609083529-2b129d95495e
 	k8s.io/api v0.32.3
 	k8s.io/apimachinery v0.32.3
+	k8s.io/utils v0.0.0-20250321185631-1f6e0b77f77e
```
[the remark this answers is not in the document]
Comment (apahim) [reply api/go.mod]: k8s.io/utils doesn't publish semantic versioned tags - only pseudo-versions

File: api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml (hunk 21 above)
[the remark this answers is not in the document]
Comment (muraee) [reply api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml:4859]: @apahim mind sharing why not do this?

File: api/go.mod (hunk 22 above)
[the remark this answers is not in the document]
Comment (apahim) [reply api/go.mod]: ok

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-TechPreviewNoUpgrade.crd.yaml [hunk 23]
```diff
@@ -5414,22 +5429,213 @@ spec:
...
+                            invalid examples: "us1" (no hyphen), "us-central" (no trailing digits), "us-central1-a" (zone suffix)
                           For a full list of valid regions, see: https://cloud.google.com/compute/docs/regions-zones.
                         maxLength: 63
-                        pattern: ^[a-z]+-[a-z0-9]+[0-9]$
+                        minLength: 1
+                        pattern: ^[a-z]+(-[a-z0-9]+)+[0-9]+$
                         type: string
                         x-kubernetes-validations:
                         - message: Region is immutable
                           rule: self == oldSelf
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-TechPreviewNoUpgrade.crd.yaml:5443]: Suggested Pattern is WRONG ❌

  The pattern suggested  ^[a-z]+(?:-[a-z0-9]+)*-[a-z]+[0-9]$ FAILS to match valid GCP regions:
  - ✗ Rejects europe-west10
  - ✗ Rejects europe-west12


I will fix it with ^[a-z]+-[a-z]+[0-9]+$ as it works for all regions

File: api/hypershift/v1beta1/gcp_validation_test.go [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+
+import (
+	"testing"
+
+	"k8s.io/utils/ptr"
+)
+
+func TestGCPResourceLabel(t *testing.T) {
```
Comment (muraee) [inline api/hypershift/v1beta1/gcp_validation_test.go]: those tests don't seem to be doing anything

File: api/hypershift/v1beta1/gcp.go [hunk 25]
```diff
@@ -4,14 +4,50 @@ package v1beta1
...
+	// value is the value part of the label. A label value can have a maximum of 63 characters.
+	// Empty values are allowed by GCP. If non-empty, it must start with a lowercase letter and
+	// may contain only lowercase letters, digits, underscores, or hyphens.
+	// See https://cloud.google.com/compute/docs/labeling-resources for Compute Engine label requirements.
+	//
+	// +optional
+	// +kubebuilder:validation:MinLength=0
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:Pattern=`^$|^[a-z][a-z0-9_-]{0,62}$`
+	Value *string `json:"value,omitempty"`
```
Comment (muraee) [inline api/hypershift/v1beta1/gcp.go:49]: since this is optional with omitempty, it doesn't need to be a pointer

File: api/hypershift/v1beta1/gcp.go [hunk 26]
```diff
@@ -81,4 +123,142 @@ type GCPPlatformSpec struct {
...
+
+	// resourceLabels are applied to all GCP resources created for the cluster.
+	// Labels are key-value pairs used for organizing and managing GCP resources.
+	// Changes to this field will be propagated in-place to GCP resources where supported.
+	// GCP supports a maximum of 64 labels per resource. HyperShift reserves approximately 4 labels for system use.
+	// For GCP labeling guidance, see https://cloud.google.com/compute/docs/labeling-resources
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=60
+	ResourceLabels []GCPResourceLabel `json:"resourceLabels,omitempty"`
```
Comment (muraee) [inline api/hypershift/v1beta1/gcp.go:138]: any reason not to mark this as a map type `+listType=map` to ensure unique keys? I believe coderrabit suggested this already,

File: api/hypershift/v1beta1/gcp.go (hunk 25 above)
Comment (apahim) [reply api/hypershift/v1beta1/gcp.go:49]: right, good point

File: api/hypershift/v1beta1/gcp_validation_test.go (hunk 24 above)
Comment (apahim) [reply api/hypershift/v1beta1/gcp_validation_test.go]: fixing

File: api/hypershift/v1beta1/gcp.go (hunk 26 above)
Comment (apahim) [reply api/hypershift/v1beta1/gcp.go:138]: done

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml [hunk 27]
```diff
@@ -5690,35 +5706,224 @@ spec:
...
+                                  Typically obtained from the output of `hypershift infra create gcp` which creates
+                                  the required service accounts with appropriate IAM roles and WIF bindings.
+                                maxLength: 100
+                                minLength: 37
+                                pattern: ^[a-z][a-z0-9-]{4,28}[a-z0-9]@[a-z][a-z0-9-]{4,28}[a-z0-9]\.iam\.gserviceaccount\.com$
+                                type: string
+                                x-kubernetes-validations:
+                                - message: NodePool is immutable
+                                  rule: self == oldSelf
+                            required:
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml:5899]: ok

File: cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml [hunk 28]
```diff
@@ -5690,35 +5706,224 @@ spec:
...
+                              type: string
+                          required:
+                          - key
+                          type: object
+                        maxItems: 60
+                        type: array
+                        x-kubernetes-list-map-keys:
+                        - key
+                        x-kubernetes-list-type: map
+                      workloadIdentity:
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/install/assets/hypershift-operator/zz_generated.crd-manifests/hostedcontrolplanes-CustomNoUpgrade.crd.yaml:5782]: ok

File: api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml [hunk 29]
```diff
@@ -4771,35 +4787,225 @@ spec:
...
+                                  This is a user-provided value referencing a pre-created Google Service Account.
+                                  Typically obtained from the output of `hypershift infra create gcp` which creates
+                                  the required service accounts with appropriate IAM roles and WIF bindings.
+                                maxLength: 100
+                                minLength: 38
+                                pattern: ^[a-z][a-z0-9-]{4,28}[a-z0-9]@[a-z][a-z0-9-]{4,28}[a-z0-9]\.iam\.gserviceaccount\.com$
+                                type: string
+                                x-kubernetes-validations:
+                                - message: NodePool is immutable
+                                  rule: self == oldSelf
```
[the remark this answers is not in the document]
Comment (apahim) [reply api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml:4979]: in the previous round you asked me to change it to 38, you have to stop changing your mind.

File: api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml [hunk 30]
```diff
@@ -4771,35 +4787,225 @@ spec:
...
+                              type: string
+                          required:
+                          - key
+                          type: object
+                        maxItems: 60
+                        type: array
+                        x-kubernetes-list-map-keys:
+                        - key
+                        x-kubernetes-list-type: map
+                      workloadIdentity:
```
[the remark this answers is not in the document]
Comment (apahim) [reply api/hypershift/v1beta1/zz_generated.featuregated-crd-manifests/hostedclusters.hypershift.openshift.io/GCPPlatform.yaml:4863]: did this on your request, no  new changes.

File: cmd/cluster/gcp/create.go [hunk 31]
```diff
@@ -55,6 +90,27 @@ func (o *RawCreateOptions) Validate(_ context.Context, _ *core.CreateOptions) (c
...
+	}
+	if err := util.ValidateRequiredOption(flagWorkloadIdentityProviderID, o.WorkloadIdentityProviderID); err != nil {
+		return nil, err
+	}
+	if err := util.ValidateRequiredOption(flagNodePoolServiceAccount, o.NodePoolServiceAccount); err != nil {
+		return nil, err
+	}
+	if err := util.ValidateRequiredOption(flagControlPlaneServiceAccount, o.ControlPlaneServiceAccount); err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (apahim) [reply cmd/cluster/gcp/create.go:113]: Are are going to support Workload Identity only at this stage. If we ever introduce Service Accounts for GCP, we make this optional.
